## **PROJECT ARTIFICIAL NEURAL NETWORK (CONVOLUTIONAL NEURAL NETWORK)**

Name : Satria Wiro Agung

NIM : 2101662133

Class : LB08


## Convolutional Neural Network with Fashion.Mnist dataset

Training Data : 48000 images

Validation Data : 12000 images

Testing Data : 10000 images




In [0]:
from keras.models import Sequential
from keras.layers import Conv2D, MaxPooling2D, Dense, Flatten, Dropout
from keras.optimizers import Adam
from keras.callbacks import TensorBoard
from keras.datasets import fashion_mnist
from keras.utils import np_utils
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from keras.optimizers import SGD

## **LOAD DATA AND PREPROCESSING**

In [0]:
((trainDatas, trainLabels), (testDatas, testLabels)) = fashion_mnist.load_data()
trainDatas = trainDatas.astype("float32") / 255.0
testDatas = testDatas.astype("float32") / 255.0

# One-hot Encoding untuk trainLabels dan testLabels
trainLabels = np_utils.to_categorical(trainLabels, 10)
testLabels = np_utils.to_categorical(testLabels, 10)

# Memisahkan training dataset menjadi trainData dan validateDatas

trainDatas, validateDatas, trainLabels, validateLabels = train_test_split(
    trainDatas, trainLabels, test_size = 0.2, random_state = 42
)


# Reshape Data
img_rows = 28
img_cols = 28
batch_size = 500
img_shape = (img_rows, img_cols, 1)

trainDatas = trainDatas.reshape(trainDatas.shape[0], *img_shape)
testDatas = testDatas.reshape(testDatas.shape[0], *img_shape)
validateDatas = validateDatas.reshape(validateDatas.shape[0], *img_shape)
print(trainDatas.shape[3])

Data dan label untuk training dan testing diimport dari dataset Fashion_Mnist. Setiap data/image berukuran 28x28 pixels dengan nilai setiap pixel 0 - 255. Untuk dapat digunakan dalam komputasi dan juga untuk mengurangi kompleksitas kalkulasi maka value dari setiap pixel pada train data dan test data harus diubah ke float dan dibagi 255 untuk mengubah nilai pixel diantara 0 - 1. 

Kemudian Label akan diubah menjadi One Hot Encoding. Contoh untuk label "T-Shirt" maka One Hot Encodingnya = [1,0,0,0,0,0,0,0,0,0], untuk label "Trouser" maka = [0,1,0,0,0,0,0,0,0,0].
One Hot Encoding digunakan untuk merepresentasikan data yang berkategori.

Train Data kemudian akan dibagi menjadi 2 yaitu Train Data(80%) dan Validation Data(20%). Validation bertujuan untuk mencegah overfitting.

Train Data, Test Data dan Validation Data akan direshape menjadi 28 x 28 greyscale.


## **CREATE CNN MODEL**

In [0]:
# CNN MODEL
name = 'CNN_1Layer'
cnn_model_1Layer = Sequential([
    Conv2D(32, kernel_size = 3, activation = 'relu',input_shape = img_shape, name = 'Conv2D-1'),
    MaxPooling2D(pool_size = 2, name='MaxPool'),
    Dropout(0.2, name='Dropout'),
    Flatten(name='Flatten'),
    
    Dense(32, activation = 'relu', name='Dense'),
    Dense(10, activation = 'softmax', name='Output')
],name = name)

name = 'CNN_2Layer'
cnn_model_2Layer = Sequential([
    Conv2D(32, kernel_size = 3, activation = 'relu', input_shape = img_shape, name='Conv2D-1'),
    MaxPooling2D(pool_size = 2, name = 'MaxPool'),
    Dropout(0.2, name='Dropout-1'),
    
    Conv2D(64,kernel_size = 3, activation = 'relu', name='Conv2D-2'),
    Dropout(0.25, name='Dropout-2'),
    Flatten(name='Flatten'),
    
    Dense(64, activation = 'relu', name='Dense'),
    Dense(10, activation = 'softmax', name='Output')
],name = name)

name='CNN_3Layer'
cnn_model_3Layer = Sequential([
    Conv2D(32, kernel_size=3, activation='relu', input_shape=img_shape, kernel_initializer='he_normal', name='Conv2D-1'),
    MaxPooling2D(pool_size=2, name='MaxPool'),
    Dropout(0.25, name='Dropout-1'),
    
    Conv2D(64, kernel_size=3, activation='relu', name='Conv2D-2'),
    Dropout(0.25, name='Dropout-2'),
    
    Conv2D(128, kernel_size=3, activation='relu', name='Conv2D-3'),
    Dropout(0.4, name='Dropout-3'),
    Flatten(name='flatten'),
    
    Dense(128, activation='relu', name='Dense'),
    Dropout(0.4, name='Dropout'),
    Dense(10, activation='softmax', name='Output')
], name=name)

cnn_models = [cnn_model_1Layer,cnn_model_2Layer,cnn_model_3Layer]
for model in cnn_models:
  model.summary();

Model: "CNN_1Layer"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
Conv2D-1 (Conv2D)            (None, 26, 26, 32)        320       
_________________________________________________________________
MaxPool (MaxPooling2D)       (None, 13, 13, 32)        0         
_________________________________________________________________
Dropout (Dropout)            (None, 13, 13, 32)        0         
_________________________________________________________________
Flatten (Flatten)            (None, 5408)              0         
_________________________________________________________________
Dense (Dense)                (None, 32)                173088    
_________________________________________________________________
Output (Dense)               (None, 10)                330       
Total params: 173,738
Trainable params: 173,738
Non-trainable params: 0
__________________________________________________

3 Model CNN yang akan digunakan : 


**Model 1 : 1 Convolutional Layer**

Convolution Layer dengan 32 filters/nodes menggunakan activition function relu. Setelah itu akan MaxPooling(Size = 2) untuk mengekstrak pixel/feature yang paling dominan dari hasil convolution layer. Kemudian Dropout(20%) akan dilakukan untuk mencegah/mengurangi overfitting. Terakhir hasil dari proses diatas akan diflatten (diubah menjadi sebuah vektor).


Dense Layer dengan 32 node , Activition Function : ReLu


Dense Layer dengan 10 node(menandakan 10 label dari data Fashion.Mnist). Activition Function : Softmax.


**Model 2 : 2 Convolutional Layer**

Convolutional Layer ke 1  : 32 filter, kernel size = 3, Activition : Relu,
Kemudian Max Pooling dengan size = 2, Dropout = 25 persen.

Convolutional Layer ke 2  : 64 filter, kernel size = 3, Activition : Relu,
Kemudian Dropout = 25 persen dan hasil di Flatten.

Dense Layer 1 : 64 Node, Activition = Relu.

Dense Layer 2 (Output Layer)  : 10 Node , Activition = Softmax


**Model 3 : 3 Convolutional Layer**

Convolutional Layer ke 1  : 32 filter, kernel size = 3, Activition : Relu,
Kemudian Max Pooling dengan size = 2, Dropout = 25 persen.

Convolutional Layer ke 2  : 64 filter, kernel size = 3, Activition : Relu,
Kemudian Dropout = 25 persen.

Convolutional Layer ke 3  : 128 filter, kernel size = 3, Activition : Relu,
Kemudian Dropout = 25 persen dan hasil di Flatten.

Dense Layer 1 : 64 Node, Activition = Relu. Kemudian Dropout 40 persen.

Dense Layer 2 (Output Layer)  : 10 Node , Activition = Softmax









## **TRAIN CNN MODEL**

Loss function yang digunakan adalah Categorical Crossentropy.
Optimizer yang digunakan adalah Adam.
Jumlah Epoch : 50
Batch Size : 500

In [0]:
# Train Model dan Hasil disimpan dalam dictionary
result_dict = {}

for model in cnn_models:
    model.compile(
        loss='categorical_crossentropy',
        optimizer=Adam(),
        metrics=['accuracy']
    )
    
    result = model.fit(
        trainDatas, trainLabels,
        batch_size=batch_size,
        epochs=50, verbose=1,
        validation_data=(validateDatas, validateLabels)
    )
    
    result_dict[model.name] = result



Instructions for updating:
Use tf.where in 2.0, which has the same broadcast rule as np.where



Train on 48000 samples, validate on 12000 samples
Epoch 1/50





48000/48000 [==============================] - 22s 463us/step - loss: 0.7615 - acc: 0.7526 - val_loss: 0.4775 - val_acc: 0.8390
Epoch 2/50
48000/48000 [==============================] - 20s 424us/step - loss: 0.4385 - acc: 0.8471 - val_loss: 0.3952 - val_acc: 0.8627
Epoch 3/50
48000/48000 [==============================] - 20s 422us/step - loss: 0.3883 - acc: 0.8639 - val_loss: 0.3618 - val_acc: 0.8754
Epoch 4/50
48000/48000 [==============================] - 20s 415us/step - loss: 0.3603 - acc: 0.8733 - val_loss: 0.3474 - val_acc: 0.8782
Epoch 5/50
48000/48000 [==============================] - 20s 420us/step - loss: 0.3375 - acc: 0.8820 - val_loss: 0.3257 - val_acc: 0.8861
Epoch 6/50
48000/48000 [==============================] - 20s 418us/step - loss: 0.3209 - acc: 0.8873 - val_loss: 0.3205 - val_acc: 0.8858
Epoch 7/50
4

## **TESTING AND RESULTS**

In [0]:
count = 0;
for model in cnn_models:
  count = count + 1
  print("=============================================================")
  print("|            MODEL ",count," (",count," CONVOLUTIONAL LAYER)            |")
  print("=============================================================")
  score,acc = model.evaluate(testDatas,testLabels)
  print("-------------------------------------------------------------")
  print("Test Score    : " , score)
  print("Test Accuracy : " , acc)
  print("-------------------------------------------------------------")
  print("=============================================================")
  print("\n")
 

|            MODEL  1  ( 1  CONVOLUTIONAL LAYER)            |
10000/10000 [==============================] - 2s 181us/step
-------------------------------------------------------------
Test Score    :  0.26603957515954973
Test Accuracy :  0.912
-------------------------------------------------------------


|            MODEL  2  ( 2  CONVOLUTIONAL LAYER)            |
10000/10000 [==============================] - 3s 285us/step
-------------------------------------------------------------
Test Score    :  0.24966826938390732
Test Accuracy :  0.9253
-------------------------------------------------------------


|            MODEL  3  ( 3  CONVOLUTIONAL LAYER)            |
10000/10000 [==============================] - 6s 553us/step
-------------------------------------------------------------
Test Score    :  0.26061647060588
Test Accuracy :  0.926
-------------------------------------------------------------




Dari diatas diketahui model dengan akurasi tertinggi adalah Model 3 (3 Convolutional Layer).
Urutan model berdasarkan akurasi : 
1. Model 3 (0.926)
2. Model 2 (0.9253)
3. Model 1 (0.912)

Test Score adalah hasil evaluasi dari loss function. Semakin kecil loss semakin kecil prediction errornya. Jika berdasarkan test score maka yang prediksinya paling sedikit errornya adalah model ke 2.

Dari hasil diatas dapat disimpulkan : 

- Model 1 dengan 1 convolutional layer memiliki akurasi dan loss yang lebih buruk dari pada model 2 dan model 3.
- Model 3 dengan 3 convolutional layer memilki akurasi paling bagus daripada model 1 dan model 2. Tetapi memiliki loss yang lebih tinggi dari pada model 2.
- Model 2 dengan 2 convolutional layer memiliki loss paling rendah daripada yang lain. Dari segi akurasi, model 2 hanya berbeda 0.001 (0.1 persen) dengan model 1 yang memiliki akurasi tertinggi.
- Secara keseluruhan, Model 2 memiliki keseimbangan/rata-rata loss dan accuracy paling bagus. Karena akurasi model 2 hanya beda sekitar 0.1 persen dengan model 1 (model dengan akurasi paling tinggi) dan model 2 mempunyai loss (kesalahan prediksi) yang paling rendah.